File: dsa-practice-ram/dsa-practice-ram/Arrays/StockBuySell.cs
namespace DsaPractice.Arrays
{
    internal class StockBuySell
    {
        public void FindStockMaxProfit()
        {
            int[] array = { 7, 1, 5, 3, 6, 4 };

            int maxProfit = int.MinValue;
            int min = array[0];

            (int buy, int sell) result = default;

            for (int i = 1; i < array.Length; i++)
            {
                if (array[i] - min > maxProfit)
                {
                    maxProfit = array[i] - min;
                    result = (min, array[i]);
                }
                min = System.Math.Min(min, array[i]);
            }

            if (maxProfit > 0)
                Console.WriteLine($"Maximum profit can be obtained if the stock is bought on {result.buy} and sold on {result.sell}");
            else
                Console.WriteLine("No profit can be made from the given stock prices.");
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/stock_buy_sell.py
"""Stock buy/sell maximum profit."""

INT_MIN = -2 ** 31

# C# StockBuySell.FindStockMaxProfit

def find_stock_max_profit(prices=None):
    """Return the best buy/sell prices and profit if positive.

    Problem: Find the maximum profit from a single buy-sell transaction.
    Approach: Track the minimum price so far and best profit in one pass.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if prices is None:
        prices = [7, 1, 5, 3, 6, 4]

    if len(prices) < 2:
        return None

    max_profit = INT_MIN
    min_price = prices[0]
    result = (min_price, min_price)

    for i in range(1, len(prices)):
        if prices[i] - min_price > max_profit:
            max_profit = prices[i] - min_price
            result = (min_price, prices[i])
        if prices[i] < min_price:
            min_price = prices[i]

    if max_profit > 0:
        return (result[0], result[1], max_profit)
    return None
